flink-runtime/src/main/java/org/apache/flink/runtime/operators/BatchTask.java

```diff
@@ -893,21 +896,34 @@ protected void resetAllInputs() throws Exception {
 					}
 				} else {
 					// close the async barrier if there is one
-					if (this.tempBarriers[i] != null) {
-						this.tempBarriers[i].close();
-						this.tempBarriers[i] = null;
-					}
+					List<MemorySegment> allocated = tempBarriers[i] == null ? emptyList() :
+						tempBarriers[i].closeAndGetLeftoverMemory();
+					tempBarriers[i] = null;
 
-					// recreate the local strategy
-					initInputLocalStrategy(i);
+					try {
+						initInputLocalStrategy(i);
 
-					if (this.inputIsAsyncMaterialized[i]) {
-						final int pages = this.materializationMemory[i];
-						@SuppressWarnings({ "unchecked", "rawtypes" })
-						TempBarrier<?> barrier = new TempBarrier(this, getInput(i), this.inputSerializers[i], memMan, ioMan, pages);
-						barrier.startReading();
-						this.tempBarriers[i] = barrier;
-						this.inputs[i] = null;
+						if (this.inputIsAsyncMaterialized[i]) {
+							final int pages = this.materializationMemory[i];
+							if (pages > allocated.size()) {
```

Comment:
This condition seems the wrong way around.

Reply:
You are right, thanks!

I changed it to just
```Preconditions.checkState(allocated.size() <= pages);```

(`pages` should never change as you pointed out, but some of the allocated segments might have been consumed, so it's `<=` and not `==`)